src/libtiled/wangset.cpp

```diff
@@ -288,11 +473,62 @@ bool WangSet::wangIdIsValid(WangId wangId) const
         if (wangId.edgeColor(i) > mEdgeColors
                 || wangId.cornerColor(i) > mCornerColors)
             return false;
+
+        if (mEdgeColors <= 1)
+            if (wangId.edgeColor(i))
+                return false;
+
+        if (mCornerColors <= 1)
+            if (wangId.cornerColor(i))
+                return false;
     }
 
     return true;
 }
 
+bool WangSet::wangIdIsUsed(WangId wangId) const
+{
+    unsigned mask = 0;
+
+    if (mEdgeColors > 1)
+        mask |= 0x0f0f0f0f;
+    if (mCornerColors > 1)
+        mask |= 0xf0f0f0f0;
+
+    for (WangId usedId : mWangIdToWangTile.keys()) {
+        if ((usedId & mask) == (wangId & mask))
+            return true;
+    }
+
+    return false;
+}
+
+WangId WangSet::templateWangIdAt(unsigned n) const
+{
+    unsigned wangId = 0;
+    //number of permutations of a corner and edge together.
+    int cornerEdgePermutations = mEdgeColors * mCornerColors;
+
+    for (int i = 7; i >= 0; --i) {
+        //this is the number of permutations possible bellow this point in the wangId
+        int bellowPermutations = qPow(cornerEdgePermutations, i/2) * ((i&1)? mEdgeColors : 1);
+        int value = n / bellowPermutations;
+        n -= value * bellowPermutations;
+
+        wangId |= value << i * 4;
+    }
+
+    //before this is like a base 10 range (0 - 9) where we want (1 - 10) for each digit
+    wangId += 0x11111111;
+    //If edges/corners don't have variations then those spots should be wild.
```

Comment:
The math above treats each digit like a traditional base n number, having range [0 - (n-1)]. Though wangId digits range from [1 - n]. So adding 1 to each digit converts between the systems. Though I suppose this is an area where having 0xf be the wild card may look a bit nicer.